modules/ctl/src/main/scala/free/ctl.scala

```diff
@@ -0,0 +1,124 @@
+        case Server.Remote(Host.Network(h), u)  => u.foldRight(h)(_ + "@" + _)
+      }
+  }
+  object Server {
+    case object Local extends Server
+    case class  Remote(host: Host, user: Option[String]) extends Server
+  }
+
+
+  /** ADT for output log levels. */
+  sealed trait Level
+  object Level {
+    case object Info  extends Level
+    case object Warn  extends Level
+    case object Error extends Level
+    case object Shell extends Level
+  }
+
+  /** ADT of low-level control operations. */
+  sealed trait CtlOp[A]
+  object CtlOp {
+    case class  Exit[A](exitCode: Int)                              extends CtlOp[A]
+    case object GetConfig                                           extends CtlOp[Config]
+    case class  Gosub[A](level: Level, msg: String, fa: CtlIO[A])   extends CtlOp[A]
+    case class  Shell(remote: Boolean, cmd: String \/ List[String]) extends CtlOp[Output]
+  }
+
+  /** Combinator that applies a check to command output. */
+  def require[A](shell: CtlIO[Output])(f: PartialFunction[Output, A]): CtlIO[A] =
+    shell.flatMap { o =>
+      f.lift(o) match {
+        case None    =>
+          o.lines.traverse(error(_))      *>
+          error(s"exited (${o.exitCode})") *>
+          exit[A](o.exitCode)
+        case Some(a) => a.point[CtlIO]
+      }
+    }
+
+  /** Syntax to allow checks on shell command output. */
+  implicit class CtlIOOps[A](fa: CtlIO[A]) {
+    def require[B](f: PartialFunction[Output, B])(implicit ev: A =:= Output): CtlIO[B] =
+      ctl.require(fa.map(ev))(f)
+  }
+
+  /** Free monad over CtlOp. */
+  type CtlIO[A] = Free[CtlOp, A]
+
+  // Constructors
+
+  def serverHostName: CtlIO[String] =
+    server.flatMap {
+      case Server.Local => "localhost".point[CtlIO]
+      case Server.Remote(Host.Network(name), _) => name.point[CtlIO]
+      case Server.Remote(Host.Machine(name), _) =>
+        shell("docker-machine", "ip", name).require {
+          case Output(0, s :: Nil) => s
+        }
+    }
+
+  def shell(c: String): CtlIO[Output] =
+    Free.liftF(CtlOp.Shell(false, c.left))
+
+  def shell(c: String, cs: String*): CtlIO[Output] =
+    Free.liftF(CtlOp.Shell(false, (c :: cs.toList).right))
+
+  def remote(c: String): CtlIO[Output] =
+    Free.liftF(CtlOp.Shell(true, c.left))
+
+  def remote(c: String, cs: String*): CtlIO[Output] =
+    Free.liftF(CtlOp.Shell(true, (c :: cs.toList).right))
+
+  def exit[A](exitCode: Int): CtlIO[A] =
+    Free.liftF(CtlOp.Exit[A](exitCode))
+
+  val config: CtlIO[Config] =
+    Free.liftF(CtlOp.GetConfig)
+
+  val server: CtlIO[Server] =
+    config.map(_.server)
+
+  val isRemote: CtlIO[Boolean] =
+    server.map {
+      case Server.Remote(_, _) => true
+      case Server.Local        => false
+    }
+
+  def gosub[A](msg: String)(fa: CtlIO[A]): CtlIO[A] =
+    Free.liftF(CtlOp.Gosub(Level.Info, msg, fa))
+
+  def logMessage(level: Level, msg: String): CtlIO[Unit] =
+    Free.liftF(CtlOp.Gosub(level, msg, ().point[CtlIO]))
```

Comment:
Ok.

When you use `Free` you construct a program that is a value, and interpret it later. It's exactly the same idea as using `IO` or `Task` but instead of being able to do absolutely anything you can restrict operations to a particular domain, "shell scripting" in this case, by providing a small number of primitive operations and building things up. That's what we're doing here. Everything you can do in this language is defined in `ctl.scala`:

```scala
/** ADT of low-level control operations. */
sealed trait CtlOp[A]
object CtlOp {
  case class  Exit[A](exitCode: Int)                              extends CtlOp[A]
  case object GetConfig                                           extends CtlOp[Config]
  case class  Gosub[A](level: Level, msg: String, fa: CtlIO[A])   extends CtlOp[A]
  case class  Shell(remote: Boolean, cmd: String \/ List[String]) extends CtlOp[Output]
}
```

So all the deployment code, etc., is define in terms of these 4 operations. Logging is defined in terms of a `Gosub` operation that says "log the execution of this sub-program". And the interpreter `interp.scala` implements this by building a tree of log statements that shows the logical execution. The `logMessage` operation above uses an empty sub-program, so the logged message has no child messages.

And @cquiroz yes, exactly, you can change the interpreter to store the logs to a file or a JSON object or whatever. The *meaning* of these operations is entirely up to the interpreter.